### PR TITLE
Darcy flow

### PR DESCRIPTION
**Pull request by danieldouglas92**
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: include/aspect/simulator/assemblers/advection.h [hunk 1]
```diff
@@ -24,6 +24,7 @@
 
 #include <aspect/simulator/assemblers/interface.h>
 #include <aspect/simulator_access.h>
+#include <aspect/melt.h>
```
Comment (naliboff) [inline include/aspect/simulator/assemblers/advection.h:27]: Out of curiosity, why is this needed?

File: source/simulator/assemblers/advection.cc [hunk 2]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+
+
+
+
+
+
+
+
+
+
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:503]: Remove some of the extra empty lines above (only need three between functions) :)

File: source/simulator/assemblers/advection.cc [hunk 3]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+    void
+    DarcySystem<dim>::create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+    {
+      MeltHandler<dim>::create_material_model_outputs(outputs);
+    }
+
+    template <int dim>
+    void
+    DarcySystem<dim>::execute (internal::Assembly::Scratch::ScratchBase<dim>   &scratch_base,
+                                   internal::Assembly::CopyData::CopyDataBase<dim> &data_base) const
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:514]: The indentation looks off here, were you able to successfully run `make indent`?

File: source/simulator/assemblers/advection.cc [hunk 4]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+      const Introspection<dim> &introspection = this->introspection();
+      const FiniteElement<dim> &fe = this->get_fe();
+
+      const typename Simulator<dim>::AdvectionField advection_field = *scratch.advection_field;
+      if (advection_field.advection_method(introspection)
+          != Parameters<dim>::AdvectionFieldMethod::fem_darcy_field)
+        return;      
+
+      
+
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:528]: Remove this empty line and the one above.

File: source/simulator/assemblers/advection.cc [hunk 5]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+               :
+               scratch.old_field_values[q]);
+
+
+          const double rho_s = scratch.material_model_outputs.densities[q];
+          const double rho_f = melt_outputs->fluid_densities[q];
+          const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q));
+          const double K_D = melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q];
+          Tensor<1,dim> current_u = scratch.current_velocity_values[q];
+          Tensor<1,dim> current_u_f = current_u - K_D * (rho_s * gravity - rho_f * gravity);
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:602]: The Darcy equation we looked at earlier has the gradient of pressure instead of`rho_s * g`. Is using `rho_s * g` equivalent to assuming all pressure variations are hydrostatic? If so, should we consider adding in an option to use the full pressure gradient? As currently constructed, this formulation would not have dynamic pressure variations included, but perhaps we don't need them as `current_u ` will include these effects to some degree? 

@jdannberg - thoughts on this?

File: include/aspect/simulator/assemblers/advection.h (hunk 1 above)
Comment (naliboff) [reply include/aspect/simulator/assemblers/advection.h:27]: Nevermind, I see below it is used for accessing permeabilities.

File: source/simulator/assemblers/advection.cc [hunk 6]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+          {
+            return 0.;
+          }  
+
+
+      const MaterialModel::MeltOutputs<dim> *melt_outputs = scratch.material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
+      for (unsigned int q=0; q < n_q_points; ++q)
+        {
+
+          const double K_D = melt_outputs->permeabilities[q] / melt_outputs->fluid_viscosities[q];
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:711]: Eventually add documentation here about these variables, which are the key component of the of the new advection method.

File: source/simulator/assemblers/advection.cc [hunk 7]
```diff
@@ -468,6 +468,307 @@ namespace aspect
...
+
+          const double dreaction_term_dt = (this->get_old_timestep() == 0) ? 0.0 :
+                                            scratch.material_model_outputs.reaction_terms[q][advection_field.compositional_variable]
+                                            / this->get_old_timestep();
+
+          residuals[q] = std::abs(dField_dt + u_grad_field - dreaction_term_dt);
+        }
+      return residuals;
+    }
+
```
Comment (naliboff) [inline source/simulator/assemblers/advection.cc:735]: Delete empty lines below

File: source/simulator/assembly.cc [hunk 8]
```diff
@@ -206,6 +206,20 @@ namespace aspect
...
 
+
+
+
+    // add the darcy assemblers if we have fields that use this method
+    if (std::find(parameters.compositional_field_methods.begin(), parameters.compositional_field_methods.end(),
+                  Parameters<dim>::AdvectionFieldMethod::fem_darcy_field) != parameters.compositional_field_methods.end())
+      assemblers->advection_system.push_back(
+        std::make_unique<aspect::Assemblers::DarcySystem<dim>>());
+
```
Comment (naliboff) [inline source/simulator/assembly.cc:217]: Remove extra lines

File: source/simulator/melt.cc [hunk 9]
```diff
@@ -410,6 +410,8 @@ namespace aspect
       const double pressure_scaling = this->get_pressure_scaling();
 
       std::vector<double> reactions(n_q_points, numbers::signaling_nan<double>());
+      std::cout << n_q_points << std::endl;
```
Comment (naliboff) [inline source/simulator/melt.cc:413]: This should be removed when it is no longer needed for debugging.

File: source/simulator/parameters.cc [hunk 10]
```diff
@@ -1262,6 +1262,22 @@ namespace aspect
                          "This method can only be chosen if melt transport is active in the "
                          "model."
                          "\n"
+
+
+
+                         "\\item ``darcy field'': If a compositional field is marked with this "
```
Comment (naliboff) [inline source/simulator/parameters.cc:1268]: Remove empty lines above

File: source/simulator/parameters.cc [hunk 11]
```diff
@@ -1262,6 +1262,22 @@ namespace aspect
...
+                         "method, then its values are computed in each time step by "
+                         "advecting along the values of the previous time step using the "
+                         "melt velocity, and applying reaction rates to it. In other words, "
+                         "this corresponds to the usual notion of a composition field as "
+                         "mentioned in Section~\\ref{sec:compositional}, except that it is "
+                         "advected with the melt velocity instead of the solid velocity. "
+                         "This method can only be chosen if melt transport is active in the "
+                         "model."
+                         "\n"
+
```
Comment (naliboff) [inline source/simulator/parameters.cc:1278]: Remove empty lines below

File: tests/darcy_velocity.prm [hunk 12]
```diff
@@ -0,0 +1,157 @@
+#########################################################
+# This is a test for the new melt solver. It checks if
```
Comment (naliboff) [inline tests/darcy_velocity.prm:2]: Update the documentation here to reflect that the test is based on rising_melt_blob, but is modified to test the new Darcy flow implementation.  Likewise, update the documentation in other sections accordingly. Is there an analytical solution to this problem? If so, I think it would be good to add a benchmark example in this PR.

File: tests/darcy_velocity.prm [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+# We want to use two compositional fields, the porosity and
+# an additional field, to check if the melting rate functionality
+# has the same effect as the reaction term for the other
+# compositional fields. As the porosity field is advected by
+# a different mechanism, slight differences between the fields
+# are expected.
+subsection Compositional fields
+  set Number of fields = 3
+  set Names of fields = porosity, solid, fluid
+  set Compositional field methods = field, field, darcy field
```
Comment (naliboff) [inline tests/darcy_velocity.prm:76]: Is the porosity always advected with the solid velocity?

**Review by naliboff — CHANGES_REQUESTED** [review]
@danieldouglas92 - Nice, this is off to a great start! I have a number of minor comments and questions throughout, but to first order the structure and code looks to be in good shape. One thing to do in the short term is take care of minor formatting issues (extra spaces, indentations, etc) and updating the documentation.

**Review by jdannberg — COMMENTED** [review]
I think it is great that this works!
I left a few comments, mostly regarding things you can remove, and checks that make sure all the options chosen in the input file are consistent. But I think the core functionality works already, and everything else looks good!

File: source/simulator/assemblers/advection.cc [hunk 14]
```diff
@@ -468,6 +468,309 @@ namespace aspect
...
+    DarcySystem<dim>::compute_residual(internal::Assembly::Scratch::ScratchBase<dim> &scratch_base) const
+    {
+      internal::Assembly::Scratch::AdvectionSystem<dim> &scratch = dynamic_cast<internal::Assembly::Scratch::AdvectionSystem<dim>& > (scratch_base);
+      const typename Simulator<dim>::AdvectionField advection_field = *scratch.advection_field;
+      const unsigned int n_q_points = scratch.finite_element_values.n_quadrature_points;
+      std::vector<double> residuals(n_q_points);
+      if (advection_field.is_temperature())
+          {
+            return residuals;
+          }
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/simulator/assemblers/advection.cc]: It does crash, this was the best way I could find to keep the advection field from being the temperature field in the compute_residuals function

**Comment by jdannberg** [comment]
Just so that I do not forget this: @tjhei mentioned this morning that it would be useful to write down the equations somewhere, reference the paper they come from, and say a bit about the limitations of the method. A good place for that would be a new short section in the manual. There is already a section about models with melt transport, so it could be right after that.

File: tests/darcy_velocity.prm [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+  end
+end
+
+subsection Initial composition model
+  set Model name = function
+
+  subsection Function
+    set Variable names      = x,y
+    set Function constants  = pi=3.1415926,x0=100000,y0=50000,c=10000, A=1.0
+    set Function expression = A * exp(-((x-x0)*(x-x0)+(y-y0)*(y-y0))/(2*c*c)); A * exp(-((x-x0)*(x-x0)+(y-y0)*(y-y0))/(2*c*c)); A * exp(-((x-x0)*(x-x0)+(y-y0)*(y-y0))/(2*c*c))
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply tests/darcy_velocity.prm:116]: This is something that us actually making the test evolve strangely, so we should definitely talk about the best way to setup the initial porosity field

File: tests/darcy_velocity.prm [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+  set List of model names = initial composition
+  set Fixed composition boundary indicators = bottom, top, left, right
+end
+
+subsection Boundary temperature model
+  set List of model names = box
+
+  subsection Box
+    set Bottom temperature = 293
+    set Top temperature    = 293
```
Comment (naliboff) [inline tests/darcy_velocity.prm:80]: Also set left and right temperatures?

File: tests/darcy_velocity.prm [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+  set Compositional field methods = darcy field, field, darcy field
+  set Types of fields = porosity, generic, generic
+end
+
+subsection Boundary composition model
+  set List of model names = initial composition
+  set Fixed composition boundary indicators = bottom, top, left, right
+end
+
+subsection Boundary temperature model
```
Comment (naliboff) [inline tests/darcy_velocity.prm:75]:
```suggestion
# This model is only designed to predict fluid velocities arising from initial prescribed values in porosity.
# As such, no variations in temperature are needed and both the boundary and initial temperatures are held constant.
subsection Boundary temperature model
```

File: tests/darcy_velocity.prm [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+############### Compositional fields
+# We want to use three compositional fields, the porosity and
+# and two additional fields. The solid field will be advected with
+# the solid speed, and the fluid field will be advected as the sum of
+# the solid velocity and the vertical buoyancy the fluid has. The
+# porosity field is uniform to allow for a uniform fluid velocity.
+subsection Compositional fields
+  set Number of fields = 3
+  set Names of fields = porosity, solid, fluid
```
Comment (naliboff) [inline tests/darcy_velocity.prm:65]: Do we technically need to have the solid field, or can we just use the "background" to represent the solid phase? This might be a melt material model feature I missed.

File: tests/darcy_velocity.prm [hunk 19]
```diff
@@ -0,0 +1,148 @@
+#########################################################
+# This is a test for the new advection method that advects a
+# compositional field with the fluid velocity prescribed by
+# Darcy's Law.
+
+# The initial porosity is uniform everywhere, and in the
+# center of the model is a blob of 'fluid' and 'solid'.
+# The solid has a prescribed x velocity, and so the fluid
+# blob should be advected horizontally with the solid and
+# vertically as it rises up due to its buoyancy.
```
Comment (naliboff) [inline tests/darcy_velocity.prm:10]: I think some lines of documentation are also needed regarding the material model being used (melting_rate), which is defined in the test file `rising_melt_blob.cc`. Two-phase melt transport is turned off, but we are using the Melt interface to access and/or fill all the melt related terms. This will probably be confusing to anyone who is not familiar with the melt material model or how those work. 

After this PR is merged it would be good to add a benchmark that composites on top of the visco plastic material, and adds on all the melt-related terms.

File: tests/darcy_velocity.prm [hunk 20]
```diff
@@ -0,0 +1,148 @@
+#########################################################
+# This is a test for the new advection method that advects a
+# compositional field with the fluid velocity prescribed by
+# Darcy's Law.
+
+# The initial porosity is uniform everywhere, and in the
+# center of the model is a blob of 'fluid' and 'solid'.
+# The solid has a prescribed x velocity, and so the fluid
```
Comment (naliboff) [inline tests/darcy_velocity.prm:8]: How is the solid velocity prescribed?

File: tests/darcy_velocity.prm [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+
+
+subsection Initial temperature model
+  set Model name = function
+  subsection Function
+    set Function expression       = 293
+  end
+end
+
+subsection Initial composition model
```
Comment (naliboff) [inline tests/darcy_velocity.prm:102]: Can you add a few briefs line of documentation saying something to the effect of "The initial porosity is constant, while the solid and fluid" fields are circles of size X/Y in the center of the model domain"?

**Review by naliboff — COMMENTED** [review]
@danieldouglas92 - Nice work, this is getting close! A number of additional comments/questions, which entirely all related to documentation.

File: source/simulator/assemblers/advection.cc [hunk 22]
```diff
@@ -680,11 +611,13 @@ namespace aspect
...
           const double rho_s = scratch.material_model_outputs.densities[q];
           const double rho_f = melt_outputs->fluid_densities[q];
+          const unsigned int porosity_index = introspection.compositional_index_for_name("porosity");
+          const double porosity         = std::max(scratch.material_model_inputs.composition[q][porosity_index], 1e-10);
           const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q));
 
           const Tensor<1,dim> u = (scratch.old_velocity_values[q] +
                                    scratch.old_old_velocity_values[q]) / 2;
-          - K_D*(rho_s * gravity - rho_f*gravity);
+          - K_D*(rho_s - rho_f)*gravity/porosity;
```
[the remark this answers is not in the document]
Comment (danieldouglas92) [reply source/simulator/assemblers/advection.cc:620]: Darcy's law gives the fluid velocity as:

phi*uf = phi*us - KD * (gradP - rhof * g)
(Equation 9 in Juliane and Timo's 2016 paper)

but we divide everything by phi to isolate the fluid velocity and thats why KD is divided by the porosity

File: tests/darcy_velocity.prm (hunk 18 above)
Comment (danieldouglas92) [reply tests/darcy_velocity.prm:65]: The solid field is not needed no, I can remove it. I just had it initially to make sure that the fluid was actually moving with both the solid velocity and the buoyancy

File: source/simulator/assembly.cc [hunk 23]
```diff
@@ -206,6 +206,20 @@ namespace aspect
...
+
+
+
+
+
```
Comment (jdannberg) [inline source/simulator/assembly.cc:222]: And same here, remove the empty lines 218-222.

File: source/simulator/parameters.cc [hunk 24]
```diff
@@ -1262,6 +1262,14 @@ namespace aspect
                          "This method can only be chosen if melt transport is active in the "
                          "model."
                          "\n"
+                         "\\item ``darcy field'': If a compositional field is marked with this "
+                         "method, then its values are computed in each time step by "
+                         "advecting along the values of the previous time step using the "
+                         "fluid velocity prescribed by Darcy's Law, and applying reaction rates "
+                         "to it. In other words this corresponds to the usual notion of a composition "
+                         "field as mentioned in Section~\\ref{sec:compositional}, except that it is "
+                         "advected with the Darcy velocity instead of the solid velocity."
```
Comment (jdannberg) [inline source/simulator/parameters.cc:1271]: Please add a line that says "This advection method requires a compositional field with name and type porosity to exist in the simulation, and this field needs to be advected with the "darcy field" method as well."

Comment (jdannberg) [inline source/simulator/parameters.cc:1271]: I would also add another sentence here that says how the Darcy velocity is defined (i.e., the equation that is used to compute it, which you already have in the cc file).

File: source/simulator/parameters.cc [hunk 25]
```diff
@@ -1949,6 +1959,14 @@ namespace aspect
                    ExcMessage ("The list of names for the mapped particle property fields needs to either be empty or have a length equal to "
                                "the number of compositional fields that are interpolated from particle properties."));
 
+      if (std::find(compositional_field_methods.begin(), compositional_field_methods.end(), AdvectionFieldMethod::fem_darcy_field)
+          != compositional_field_methods.end())
+        AssertThrow (std::find(names_of_compositional_fields.begin(),
+                               names_of_compositional_fields.end(),
+                               "porosity")
+                     != names_of_compositional_fields.end(),
+                     ExcMessage ("The Darcy advection field method only works if there is a compositional field named 'porosity'"));
```
Comment (jdannberg) [inline source/simulator/parameters.cc:1968]: This is the first step, but you also need to check if the porosity field is actually advected with the darcy velocity as well.

File: tests/darcy_velocity.prm [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+# center of the model is a blob of 'fluid' and 'solid'.
+# The solid has a prescribed x velocity, and so the fluid
+# blob should be advected horizontally with the solid velocity and
+# vertically as it rises up due to its buoyancy.
+
+############### Global parameters
+
+set Dimension                              = 2
+set Start time                             = 0
+set End time                               = 5000
```
Comment (jdannberg) [inline tests/darcy_velocity.prm:27]: I would change the end time to 600, so the test is quick to run on the tester.

File: tests/darcy_velocity.prm [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+# The final part of this input file describes how many times the
+# mesh is refined and what to do with the solution once computed
+subsection Mesh refinement
+  set Initial adaptive refinement        = 0
+  set Initial global refinement          = 5
+  set Time steps between mesh refinement = 0
+end
+
+subsection Postprocess
+  set List of postprocessors = composition statistics, visualization
```
Comment (jdannberg) [inline tests/darcy_velocity.prm:133]: Looking at the output, it occurs to me that there is not actually anything that would allow us to detect if the test produces the correct results (we do not know u_f, and we only have the global integral of the compositional field, so we do not know where the blob is). I would add the "maximum depth of field" postprocessor here, which shows the largest depth at which a point is found where the field has a value of 0.5 or larger. This value should decrease over time for the fluid field.

File: tests/darcy_velocity.prm [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+  subsection Function
+    set Variable names      = x,y
+    set Function constants  = pi=3.1415926,x0=100000,y0=50000,c=12500, A_f=0.2, A_s = 1.0
+    set Function expression = A_f; if((x - x0)*(x - x0) + (y - y0)*(y - y0) <= c*c, A_s, 0); if((x - x0)*(x - x0) + (y - y0)*(y - y0) <= c*c, A_s, 0)
+  end
+end
+
+subsection Boundary composition model
+  set List of model names = initial composition
+  set Fixed composition boundary indicators = bottom, top, left, right
```
Comment (jdannberg) [inline tests/darcy_velocity.prm:103]: Does this actually work correctly?
I thought you would need to `set Allow fixed composition on outflow boundaries = true` fot the boundary conditions to be applied properly.

**Review by jdannberg — COMMENTED** [review]
Okay, this looks good to me! I think the assembler itself is ready now, and the few comments I have are mostly about formatting and about the test. 
:+1:

**Comment by jdannberg** [comment]
It looks like you also need to update the indentation. 

In addition, I am making a list of things here that would be good to address in a follow-up pull request so that we don't forget about them. Once this is merged, we can put that into an issue:

- Add a section in the manual that explains the equations, why this is useful, and what the limitations are. 
- Adapt the time step according to the Darcy velocity (source/time_stepping/convection_time_step.cc)
- Make sure we take this velocity into account when setting boundary conditions (function replace_outflow_boundary_ids in helper_functions.cc)
- would a graphical output of u_f be useful?

File: tests/darcy_velocity.prm (hunk 28 above)
Comment (danieldouglas92) [reply tests/darcy_velocity.prm:103]: it seemed like it was working but I added this for good measure

File: source/simulator/parameters.cc [hunk 29]
```diff
@@ -1961,11 +1964,18 @@ namespace aspect
...
                      ExcMessage ("The Darcy advection field method only works if there is a compositional field named 'porosity'"));
+      if ((std::find(compositional_field_methods.begin(), compositional_field_methods.end(), AdvectionFieldMethod::fem_darcy_field)
+           != compositional_field_methods.end()) && (std::find(names_of_compositional_fields.begin(), names_of_compositional_fields.end(), "porosity") != names_of_compositional_fields.end()) )
+        for (unsigned int i=0; i<n_compositional_fields; ++i)
+
+          {
+            AssertThrow ((((compositional_field_methods[i] != AdvectionFieldMethod::fem_darcy_field) &&
+                           (names_of_compositional_fields[i] == "porosity")) == false),
+                         ExcMessage ("When using the darcy field method, the porosity field must be advected with the darcy method."));
+          }
```
Comment (jdannberg) [inline source/simulator/parameters.cc:1978]: I think you can make this simpler and shorter:
```suggestion
        {
          const unsigned int porosity_idx = std::find(names_of_compositional_fields.begin(), names_of_compositional_fields.end(), "porosity");
          AssertThrow (porosity_idx != names_of_compositional_fields.end(),
                       ExcMessage ("The Darcy advection field method only works if there is a compositional field named 'porosity'"));
          AssertThrow (compositional_field_methods[porosity_idx] == AdvectionFieldMethod::fem_darcy_field,
                       ExcMessage ("When using the darcy field method, the porosity field must be advected with the darcy method."));
        }
```

File: source/simulator/parameters.cc [hunk 30]
```diff
@@ -1268,7 +1268,10 @@ namespace aspect
                          "fluid velocity prescribed by Darcy's Law, and applying reaction rates "
                          "to it. In other words this corresponds to the usual notion of a composition "
                          "field as mentioned in Section~\\ref{sec:compositional}, except that it is "
-                         "advected with the Darcy velocity instead of the solid velocity."
+                         "advected with the Darcy velocity instead of the solid velocity. This method "
+                         "requires there to be a compositional field named porosity that is advected "
+                         "the darcy field method. We calculate the fluid velocity current_u_f using "
```
Comment (jdannberg) [inline source/simulator/parameters.cc:1273]:
```suggestion
                         "using the darcy field method. We calculate the fluid velocity current_u_f using "
```

**Review by jdannberg — APPROVED** [review]
I have two small suggested changes, otherwise this looks good to me and is ready to merge from my side!

**Comment by jdannberg** [comment]
@naliboff Are you happy with the new version? You have to click on "Approve" before the pull request can be merged (that's a setting we changed in the last few days; PRs can only be merged with approving reviews).

File: source/simulator/parameters.cc (hunk 25 above)
Comment (naliboff) [reply source/simulator/parameters.cc:1969]: This is now resolved with the Assert below.

File: tests/darcy_velocity.prm [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+  set Initial adaptive refinement        = 0
+  set Initial global refinement          = 5
+  set Time steps between mesh refinement = 0
+end
+
+subsection Postprocess
+  set List of postprocessors = composition statistics, maximum depth of field
+end
+
+subsection Solver parameters
```
Comment (naliboff) [inline tests/darcy_velocity.prm:137]: This section can be deleted as it just used the defaults, but let's update that with a new PR.

**Review by naliboff — APPROVED** [review]
@danieldouglas92 - The PR looks great, thank you for all the hard work on this and excited to start testing this feature!

Self-reminder for a future PR: Re-add the melt transport parameter definition to the test prm file, along with a description of why we set it to false even though a "melt" material model is used.

File: source/simulator/parameters.cc (hunk 29 above)
Comment (danieldouglas92) [reply source/simulator/parameters.cc]: I ran into an issue when just trying to implement what you have here, it might still work but std::find returns an iterator and not an unsigned int, I was not sure how best to get around this without making the code look messier again.
